stl/inc/span

```diff
@@ -0,0 +1,679 @@
+        return rbegin();
+    }
+
+    _NODISCARD constexpr const_reverse_iterator crend() const noexcept {
+        return rend();
+    }
+
+    _NODISCARD constexpr pointer _Unchecked_begin() noexcept {
+        return _Mydata;
+    }
+
+    _NODISCARD constexpr const_pointer _Unchecked_begin() const noexcept {
+        return _Mydata;
+    }
+
+    _NODISCARD constexpr pointer _Unchecked_end() noexcept {
+        return _Mydata + size();
+    }
+
+    _NODISCARD constexpr const_pointer _Unchecked_end() const noexcept {
+        return _Mydata + size();
+    }
+
+private:
+    pointer _Mydata{nullptr};
+};
+
+// [span.tuple] Tuple interface
+template <class _Ty, size_t _Extent>
+struct tuple_size<span<_Ty, _Extent>> : integral_constant<size_t, _Extent> {
+    static_assert(dynamic_extent != _Extent, "std::span<T, dynamic_extent> is not tuple-like");
+};
+
+template <size_t _Index, class _Ty, size_t _Extent>
+struct tuple_element<_Index, span<_Ty, _Extent>> {
+    static_assert(dynamic_extent != _Extent, "std::span<T, dynamic_extent> is not tuple-like");
+    static_assert(_Index < _Extent, "Index out of bounds for a std::span of this extent");
+    using type = _Ty;
+};
+
+template <size_t _Index, class _Ty, size_t _Extent>
+_NODISCARD constexpr _Ty& get(span<_Ty, _Extent> _Span) noexcept {
+    static_assert(dynamic_extent != _Extent, "std::get<> not supported for std::span<T, dynamic_extent>");
+    static_assert(_Index < _Extent, "Index out of bounds for a std::span of this extent");
+    return _Span[_Index];
+}
+
+#if _HAS_STD_BYTE
+// [span.objectrep] Views of object representation
+// FUNCTION TEMPLATE as_bytes
+template <class _Ty, size_t _Extent>
+_NODISCARD auto as_bytes(span<_Ty, _Extent> _Sp) noexcept
+    -> span<const byte, _Extent == dynamic_extent ? dynamic_extent : sizeof(_Ty) * _Extent> {
+    return {reinterpret_cast<const byte*>(_Sp.data()), _Sp.size_bytes()};
+}
+
+// FUNCTION TEMPLATE as_writable_bytes
+template <class _Ty, size_t _Extent, enable_if_t<!is_const_v<_Ty>, int> = 0>
+_NODISCARD auto as_writable_bytes(span<_Ty, _Extent> _Sp) noexcept
+    -> span<byte, _Extent == dynamic_extent ? dynamic_extent : sizeof(_Ty) * _Extent> {
+    return {reinterpret_cast<byte*>(_Sp.data()), _Sp.size_bytes()};
+}
+#endif // _HAS_STD_BYTE
+
+// DEDUCTION GUIDES
+template <class _Ty, size_t _Extent>
+span(_Ty (&)[_Extent])->span<_Ty, _Extent>;
+
+template <class _Ty, size_t _Extent>
+span(array<_Ty, _Extent>&)->span<_Ty, _Extent>;
+
+template <class _Ty, size_t _Extent>
+span(const array<_Ty, _Extent>&)->span<const _Ty, _Extent>;
+
+template <class _Container>
+span(_Container&)->span<typename _Container::value_type>;
+
+template <class _Container>
+span(const _Container&)->span<const typename _Container::value_type>;
+
+_STD_END
+
+#pragma pop_macro("new")
+_STL_RESTORE_CLANG_WARNINGS
+#pragma warning(pop)
+#pragma pack(pop)
+#else // ^^^ _HAS_CXX20 / !_HAS_CXX20 vvv
+#pragma message("span is only available with C++20 or later")

```

Comment:
We say "class template variant" in `<variant>`, do we want to add "class template" for consistency here, or strike it there?

/cc @CaseyCarter

Reply:
IMO, strike it in `<variant>`. "function template `std::visit`" is also only available with C++20 or later, we really mean "the contents of this header are only available" rather than "this particular component" but I think "meow is only available ..." is the best error message when `<meow>` exists primarily to implement a component `std::meow` .

Reply:
I am also totally in favor of doing nothing

Reply:
I've filed #250 since we're currently inconsistent.

@miscco, I do observe that your message doesn't end with a period, whereas the other messages do.